core/src/main/scala/org/apache/spark/Accumulators.scala

```diff
@@ -320,7 +334,13 @@ private[spark] object Accumulators {
   def add(values: Map[Long, Any]): Unit = synchronized {
     for ((id, value) <- values) {
       if (originals.contains(id)) {
-        originals(id).asInstanceOf[Accumulable[Any, Any]] ++= value
+        // Since we are now storing weak references, we must check whether the underlying data
+        // is valid. 
+        originals(id).get match {
+          case Some(accum) => accum.asInstanceOf[Accumulable[Any, Any]] ++= value
+          case None => 
+            throw new IllegalAccessError("Attempted to access garbage collected Accumulator.")   
```

Comment:
The exception thrown here is caught at higher levels of the stack.  For example, DAGScheduler wraps calls to accumulator methods in a `try` block and logs any uncaught exceptions.  Have you run into a case where the current behavior causes a problem?